src/download/src/lib.rs

```diff
@@ -1,75 +1,38 @@
 //! Easy file downloading
 
-use errors::*;
-use notifications::Notification;
-use sha2::{Sha256, Digest};
-use std::cell::RefCell;
-use std::fs;
-use std::io;
-use std::env;
-use std::path::Path;
-use url::Url;
-
-pub fn download_file(url: &Url,
-                     path: &Path,
-                     hasher: Option<&mut Sha256>,
-                     notify_handler: &Fn(Notification))
-                     -> Result<()> {
-
-    let file = RefCell::new(try!(fs::File::create(&path).chain_err(
-        || "error creating file for download")));
-    let hasher = RefCell::new(hasher);
-
-    // This callback will write the download to disk and optionally
-    // hash the contents, then forward the notification up the stack
-    let handler: &Fn(Notification) -> Result<()> = &|notice| {
-        match notice {
-            Notification::DownloadDataReceived(data) => {
-                try!(io::Write::write_all(&mut *file.borrow_mut(), data)
-                     .chain_err(|| "unable to write download to disk"));
-                if let Some(ref mut h) = *hasher.borrow_mut() {
-                    h.input(data);
-                }
-            }
-            _ => ()
-        }
-
-        notify_handler(notice);
-
-        Ok(())
-    };
-
-    // Download the file
-    if env::var_os("RUSTUP_USE_HYPER").is_some() {
-        notify_handler(Notification::UsingHyper);
-         try!(self::hyper::download_file(url, handler));
-    } else {
-        notify_handler(Notification::UsingCurl);
-        try!(self::curl::download_file(url, handler));
-    }
-
-    try!(file.borrow_mut().sync_data().chain_err(|| "unable to sync download to disk"));
-
-    notify_handler(Notification::DownloadFinished);
-
-    Ok(())
+#[macro_use]
+extern crate error_chain;
+extern crate url;
+
+mod errors;
+pub use errors::*;
+
+#[derive(Debug)]
+pub enum Event<'a> {
+    /// Received the Content-Length of the to-be downloaded data.
+    DownloadContentLengthReceived(u64),
+    /// Received some data.
+    DownloadDataReceived(&'a [u8]),
 }
 
-
 /// Download via libcurl; encrypt with the native (or OpenSSl) TLS
 /// stack via libcurl
-mod curl {
-    use curl::easy::Easy;
+#[cfg(feature = "curl-mode")]
+pub mod curl {
```

Comment:
I'd probably recommend extracting this to its own file eventually, having all supported providers in one file may get a little cumbersome eventually